pkg/servicemesh/federation/server/routing.go

```diff
@@ -47,8 +51,25 @@ var (
 	armageddonTime = time.Unix(1<<62-1, 0)
 )
 
+// hashResourceName applies a sha256 on the host and truncate it to the first n bytes
+func hashResourceName(resourceName string, n int) string {
+	hash := sha256.Sum256([]byte(resourceName))
+	return hex.EncodeToString(hash[:n])
+}
+
+func shortenName(name string) string {
+	return strings.Replace(name, "-", "", -1)
+}
+
 func createResourceName(mesh string, source federationmodel.ServiceKey) string {
-	return fmt.Sprintf("federation-exports-%s-%s-%s", mesh, source.Name, source.Namespace)
+	resourceName := fmt.Sprintf("federation-exports-%s-%s-%s", mesh, source.Name, source.Namespace)
+	if !labels.IsDNS1123Label(resourceName) {
+		resourceName = fmt.Sprintf("federation-exports-%s-%s-%s", shortenName(mesh), shortenName(source.Name), shortenName(source.Namespace))
+		if !labels.IsDNS1123Label(resourceName) {
+			resourceName = fmt.Sprintf("%s-%s", resourceName[:52], hashResourceName(resourceName, 5))
```

Comment:
what's the 52 definition there ? Is that a hard-coded slice length ? It's better to add a comment with an example of that slice.
If it's a hard-coded length, we may need to define a constant variable.

Reply:
I'd change this code:
- if too long, truncate to 55 characters + add 8 characters of hash
- shorten `"federation-exports-"` to `"fed-exp-"`
- change order from: `mesh-name-namespace` to `name-namespace-mesh` so that the name is truncated last
- use constants, as @yxun suggested (e.g. `maxNameLength`, `hashLength`, and use `[:maxNameLength-hashLength]` to truncate)

Please also add a unit test for the `createResourceName` function showing examples of how the function truncates the name. 

You must also add an equivalent change to https://github.com/Maistra/istio/blob/403a6b63c620b8b7c2376f3a24b17022acbc0509/pilot/pkg/serviceregistry/federation/routing.go#L36-L38, otherwise you didn't fix the issue, as it's just going to break on the import side.